spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,491 @@
+
+# %%
+from snorkel.classification.task import Operation
+
+task_flow = [
+    Operation(name="input_op", module_name="mlp", inputs=[("_input_", "bow_features")]),
+    Operation(name="head_op", module_name="prediction_head", inputs=[("input_op", 0)]),
+]
+
+# %% [markdown]
+# With these pieces, we're ready to define a [`Task`](https://snorkel.readthedocs.io/en/redux/source/snorkel.classification.html#module-snorkel.classification.task) in Snorkel for spam classification.
+
+# %%
+from functools import partial
+from snorkel.analysis import Scorer
+from snorkel.classification import (
+    Task,
+    cross_entropy_from_outputs,
+    softmax_from_outputs,
+)
+
+spam_task = Task(
+    name="spam_task",
+    module_pool=module_pool,
+    task_flow=task_flow,
+    loss_func=partial(cross_entropy_from_outputs, "head_op"),
+    output_func=partial(softmax_from_outputs, "head_op"),
+    scorer=Scorer(metrics=["accuracy", "f1"]),
+)
+
+# %% [markdown]
+# We'll initialize a [`SnorkelClassifier`](https://snorkel.readthedocs.io/en/redux/source/snorkel.classification.html) with the `spam_task` we've created, initialize a corresponding [`Trainer`](https://snorkel.readthedocs.io/en/redux/source/snorkel.classification.training.html#module-snorkel.classification.training.trainer), and `fit` to our dataloaders!
+
+# %%
+from snorkel.classification import MultitaskClassifier, Trainer
+
+model = MultitaskClassifier([spam_task])
+trainer = Trainer(n_epochs=5, lr=1e-4, progress_bar=True)
+trainer.fit(model, [dl_train, dl_valid])
+
+# %% [markdown]
+# How well does our model do?
+
+# %%
+model.score([dl_train, dl_valid], as_dataframe=True)
+
+# %% [markdown]
+# ## 2. Perform error analysis
+#
+# In overall metrics (`f1`, `accuracy`) our model appears to perform well!
+# However, we emphasize we might actually be _more_ interested in performance for application-critical subsets, or _slices_.
+#
+# Let's perform an [`error_analysis`](https://snorkel.readthedocs.io/en/redux/source/snorkel.analysis.html#module-snorkel.analysis.error_analysis) to see where our model makes mistakes.
+# We collect the predictions from the model and visualize examples in specific error buckets.
+
+# %%
+from snorkel.analysis.error_analysis import get_label_buckets
+
+outputs = model.predict(dl_valid, return_preds=True)
+error_buckets = get_label_buckets(
+    outputs["golds"]["spam_task"], outputs["preds"]["spam_task"]
+)
+
+# %% [markdown]
+# For application purposes, we might care especially about false negatives (i.e. true label was $1$, but model predicted $0$) — for the spam task, this error mode might expose users to external links that correspond to malware!
+
+# %%
+df_valid[["text", "label"]].iloc[error_buckets[(1, 0)]].head()
+
+# %% [markdown]
+# We notice that we're mis-classifying particularly some comments with shortened urls (e.g. `bit.ly/...`) — these links could redirect us to potentially dangerous websites, and we don't want our users to click them!
```

Comment:
I made this comment before, but for monitoring, slices should be determined apriori, right? not based on what some initial model got wrong. for example, we didn't get any fns that point to known illicit sites, but we DO care about those more...that's an important slice! can we change the narrative so that we decide on slices earlier (and one of them is bit.ly)

Reply:
I agree with this— error analysis first made more sense when the focus was modeling. shifted the story a bit